### PR TITLE
Not contains via counter automata

### PR DESCRIPTION
**Pull request by vhavlena**
This PR introduces kind a lot of new stuff

- extension of `DecisionProcedure` to handle disequations and not(contains) after a stable solution is found
- Implementation of Tag-based solving of a single diseqation and first step towards handling a single not(contains). This can be turned on by `smt.str.ca_constr=true` (by default it is off)
- extension of `FormulaPreprocessor` to handle not(contains) predicates as well
- quite nice things around Parikh image: at least there is a procedure creating Parikh image formula for an NFA. It might be useful for the future
- bug fixes in axiomatization a preprocessing

### CODE REVIEW

File: src/smt/theory_str_noodler/parikh_image.cpp [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+
+        LenNode left = concat_len(diseq.get_left_side(), i - 1);
+        // add symbol <P, var, 1, 0> where var is i-th variable on left side of diseq
+        ca::AtomicSymbol lats = {1, diseq.get_left_side()[i], 1, 0};
+        left.succ.push_back(LenNode(this->symbol_var.at(lats)));
+
+        LenNode right = concat_len(diseq.get_right_side(), j-1);
+        // add symbol <P, var, 2, 0> where var is j-th variable on left side of diseq
+        ca::AtomicSymbol rats = {1, diseq.get_right_side()[j], 2, 0};
+        right.succ.push_back(LenNode(this->symbol_var.at(rats)));
```
[the remark this answers is not in the document]
Comment (vhavlena) [reply src/smt/theory_str_noodler/parikh_image.cpp:369]: The second paragraph is clear to me. Regarding the first one: You say that if x and y are the mismatch variables, I take the symbol `<P,x,1>` if `x > y`? Otherwise I take `<P,x,2>`?

[the remark this answers is not in the document]
Comment (vhavlena) [reply src/smt/theory_str_noodler/parikh_image.cpp:369]: Great, thanks!

**Comment by vhavlena** [comment]
The results before the last rebase were fine. I am running the experiments after rebase now.

File: src/smt/theory_str_noodler/aut_assignment.cpp [hunk 2]
```diff
@@ -158,4 +158,34 @@ namespace smt::noodler {
             cur_level = next_level;
         }
     }
+
+    bool AutAssignment::is_flat(const BasicTerm& t) const {
+        bool flat = true;
+
+        mata::nfa::Nfa aut = *this->at(t);
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/aut_assignment.cpp:165]: Should this be a (const) reference? Otherwise we are copying the automaton.
```suggestion
        const mata::nfa::Nfa& aut = *this->at(t);
```

File: src/smt/theory_str_noodler/aut_assignment.cpp [hunk 3]
```diff
@@ -158,4 +158,34 @@ namespace smt::noodler {
...
+
+        mata::nfa::Nfa::TarjanDiscoverCallback callback {};
+        callback.scc_discover = [&](const std::vector<mata::nfa::State>& scc, const std::vector<mata::nfa::State>& tarjan_stack) -> bool {
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/aut_assignment.cpp:168]: It is better to be explicit which variables we are capturing.
```suggestion
        callback.scc_discover = [&flat](const std::vector<mata::nfa::State>& scc, const std::vector<mata::nfa::State>& tarjan_stack) -> bool {
```

Comment (jurajsic) [reply src/smt/theory_str_noodler/aut_assignment.cpp:168]: Can you also add comment what this function does?

File: src/smt/theory_str_noodler/aut_assignment.h [hunk 4]
```diff
@@ -355,6 +355,8 @@ namespace smt::noodler {
             return complement_aut(*(this->at(t)));
         }
 
+        bool is_flat(const BasicTerm& t) const;
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/aut_assignment.h:358]: Add comment explaining what is flat automaton.

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 5]
```diff
@@ -968,12 +968,16 @@ namespace smt::noodler {
                 auto [noodler_lengths, precision] = dec_proc->get_lengths();
                 lengths = len_node_to_z3_formula(noodler_lengths);
                 lbool is_lengths_sat = check_len_sat(lengths);
-
-                // we assume that precision != LenNodePrecision::OVERAPPROX
                 
                 if (is_lengths_sat == l_true) {
                     STRACE("str", tout << "len sat " << mk_pp(lengths, m) << std::endl;);
-                    last_run_was_sat = true;
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:976]: Was this lost in rebasing?

Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:976]: Probably. It should be there?

Comment (jurajsic) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:976]: Yeah, `block_curr_len` should not be called there anymore (as it was than only to keep it in loop protection, which is not needed).

File: src/smt/theory_str_noodler/decision_procedure.h [hunk 6]
```diff
@@ -400,20 +413,20 @@ namespace smt::noodler {
...
+        // lbool replace_not_contains();
+
+        // /**
+        //  * @brief Check if it is possible to syntactically unify not contains terms. If they are included (in the sense of vectors) the 
+        //  * not(contain) is unsatisfiable.
+        //  * 
+        //  * @param prep FormulaPreprocessor
+        //  * @return l_true -> can be unified 
+        //  */
+        // lbool can_unify_not_contains(const FormulaPreprocessor& prep);
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/decision_procedure.h:429]: Remove, if not needed anymore.

Comment (vhavlena) [reply src/smt/theory_str_noodler/decision_procedure.h]: Done.

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 5 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:976]: Done

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 7]
```diff
@@ -968,12 +968,17 @@ namespace smt::noodler {
...
                 lengths = len_node_to_z3_formula(noodler_lengths);
                 lbool is_lengths_sat = check_len_sat(lengths);
-
-                // we assume that precision != LenNodePrecision::OVERAPPROX
                 
                 if (is_lengths_sat == l_true) {
                     STRACE("str", tout << "len sat " << mk_pp(lengths, m) << std::endl;);
                     last_run_was_sat = true;
+                    // save the current assignment to catch it during the loop protection
+                    block_curr_len(lengths, true, false);
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:976]: I would also remove this, only if you actually want to add this in this PR (this was replaced with the `last_run_was_sat = true` line).

File: src/smt/theory_str_noodler/parikh_image.h [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+public:
+    ParikhImageNotContTag(const ca::TagAut& ca, const std::set<ca::AtomicSymbol>& atomic_symbols) 
+        : ParikhImageDiseqTag(ca, atomic_symbols), offset_var(util::mk_noodler_var_fresh("offset_var")) { }
+
+    LenNode get_not_cont_formula(const Predicate& not_cont);
+};
+
+}
+
+#endif
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/parikh_image.h:217]:
```suggestion
#endif

```

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 7 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp]: done

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 9]
```diff
@@ -2743,7 +2746,7 @@ namespace smt::noodler {
         }
 
         for(const auto& nc : this->m_not_contains_todo_rel) {
-            app_ref nc_app(m_util_s.str.mk_contains(nc.first, nc.second), m);
+            app_ref nc_app(m.mk_not(m_util_s.str.mk_contains(nc.first, nc.second)), m);
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:2749]: This is just a fix, there should have always been `mk_not`, right?

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 10]
```diff
@@ -1716,7 +1719,7 @@ namespace smt::noodler {
             // The following axioms are redundant in the sense of completeness, but in the nested replace calls 
             // they can relate the contains predicate from the general replace (and thence the SAT solver can help a lot).
             literal cnt = mk_literal(m_util_s.str.mk_contains(s, a));
-            add_axiom({~cnt, mk_eq(v, t,false)});
+            add_axiom({~cnt, mk_literal(m.mk_not(m.mk_eq(s, a))), mk_eq(v, t,false)});
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:1722]: Why was `v` replaced with `mk_literal(m.mk_not(m.mk_eq(s, a)))`? It just looks like a bit of a random change. If it is correct, maybe just check if the comments before are still correct.

File: src/smt/theory_str_noodler/decision_procedure.cpp [hunk 11]
```diff
@@ -133,9 +133,9 @@ namespace smt::noodler {
     lbool DecisionProcedure::compute_next_solution() {
 
         // if we have a not contains, we give unknown
-        if(this->not_contains.get_predicates().size() > 0) {
-            return l_undef;
-        }
+        // if(this->not_contains.get_predicates().size() > 1) {
+        //     return l_undef;
+        // }
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/decision_procedure.cpp:138]: Remove if not needed anymore.

File: src/smt/theory_str_noodler/decision_procedure.cpp [hunk 12]
```diff
@@ -603,7 +603,7 @@ namespace smt::noodler {
     std::pair<LenNode, LenNodePrecision> DecisionProcedure::get_lengths() {
         LenNodePrecision precision = LenNodePrecision::PRECISE; // start with precise and possibly change it later
 
-        if (solution.length_sensitive_vars.empty()) {
+        if (solution.length_sensitive_vars.empty() && this->not_contains.get_predicates().size() == 0 && !this->m_params.m_ca_constr) {
             // There are no length vars (which also means no disequations nor conversions), it is not needed to create the lengths formula.
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/decision_procedure.cpp:607]: The comment should be probably updated.

File: src/smt/theory_str_noodler/decision_procedure.cpp [hunk 13]
```diff
@@ -626,6 +626,18 @@ namespace smt::noodler {
...
+
+        // get the LIA formula describing solutions for special predicates
+        conjuncts.push_back(get_formula_for_ca_diseqs());
+        auto not_cont_prec = get_formula_for_not_contains();
+        if (not_cont_prec.second == LenNodePrecision::PRECISE || not_cont_prec.second == precision) {
+            conjuncts.push_back(not_cont_prec.first);
+        } else {
+            // if we should overwrite the precision, we instead return FALSE and say that we have underapproximation
+            conjuncts.push_back(LenNode(LenFormulaType::FALSE));
+            precision = LenNodePrecision::UNDERAPPROX;
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/decision_procedure.cpp:638]: Is this correct? This will just say that it is unsat, but I think we should throw an error here instead.

File: src/smt/theory_str_noodler/parikh_image.h [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+    const std::map<Transition, BasicTerm>& get_trans_vars() const {
+        return this->trans;
+    }
+
+    virtual ~ParikhImage() { }
+};
+
+
+/**
+ * @brief Parikh image computation of Tag automaton for single disequation
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/parikh_image.h:99]: What is tag automaton?

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 9 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:2749]: Yes

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 10 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:1723]: I added a comment

File: src/smt/theory_str_noodler/decision_procedure.cpp (hunk 12 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/decision_procedure.cpp]: I updated the comment

File: src/smt/theory_str_noodler/decision_procedure.cpp (hunk 13 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/decision_procedure.cpp:633]: This basically says that we completely skip this branch of computation. If we returned `unsat` in the end, the underapproximation module outputs `unknown`. But there is still hope that we find solution in another branch.

File: src/smt/theory_str_noodler/parikh_image.h (hunk 14 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/parikh_image.h:99]: Basically it is an NFA with special symbols (tags) having some semantics in Parikh formula. The notation is still work in progress and I hope that it changes at some point.

File: src/smt/theory_str_noodler/aut_assignment.cpp (hunk 3 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/aut_assignment.cpp]: I added a comment ot `is_flat` method where it should be explained.

**Comment by vhavlena** [comment]
The results
```
# of formulae: 103405
--------------------------------------------------
tool                            ✅    ❌       time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ------  ----  ---------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-146dbc0-d95fe13  102391  1014   11437.14   0.11   0.02   1.50  62907    39484        537   409        68        0
z3-noodler-a8ca33a-d95fe13  102338  1067   12090.90   0.12   0.02   1.67  62907    39431        625   389        53        0
cvc5-1.1.2                   99774  3631   76479.19   0.77   0.00   6.77  60870    38904          2  3620         9        0
z3-4.13.0                    97694  5711  132044.66   1.35   0.01   8.11  58876    38818        210  4968       463       70
--------------------------------------------------
```
image

**Comment by vhavlena** [comment]
Main difference to the devel:
```
Benchmark str_small_rw
# of formulae: 1880
--------------------------------------------------
tool                          ✅    ❌    time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
--------------------------  ----  ----  ------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-146dbc0-d95fe13  1816    64  343.62   0.19   0.01   2.86      4     1812          4    37        23        0
z3-noodler-a8ca33a-d95fe13  1762   118  365.65   0.21   0.01   2.74      4     1758         92    17         9        0
cvc5-1.1.2                  1861    19   31.09   0.02   0.00   0.56     20     1841          2    17         0        0
z3-4.13.0                   1821    59  214.88   0.12   0.01   2.86     25     1796          0    59         0        0
--------------------------------------------------
```

File: src/smt/theory_str_noodler/decision_procedure.cpp (hunk 13 above)
Comment (jurajsic) [reply src/smt/theory_str_noodler/decision_procedure.cpp:633]: Ah, it is handled in `smt_tactic_core.cpp`, right?

File: src/smt/theory_str_noodler/expr_solver.h [hunk 15]
```diff
@@ -26,9 +26,7 @@ namespace smt::noodler {
        }
 
         lbool check_sat(expr* e, bool produce_model, model_ref &res_model);
-
-        void initialize(context& ctx, bool include_ass = true);
-
+        void initialize(context& ctx, bool include_ass);
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/expr_solver.h:29]: Why this change?

File: src/smt/theory_str_noodler/formula_preprocess.cpp [hunk 16]
```diff
@@ -1649,4 +1652,65 @@ namespace smt::noodler {
...
+            Concat right = pred.get_params()[1];
+            if(left.size() == 1 && right.size() == 1) {
+                if(this->aut_ass.is_singleton(left[0]) && this->aut_ass.is_singleton(right[0])) {
+                    if(mata::nfa::are_equivalent(*this->aut_ass.at(left[0]), *this->aut_ass.at(right[0]))) {
+                        return false;
+                    }
+                }
+            }
+            if(left.size() == 1 && right.size() == 1) {
+                if(this->aut_ass.is_singleton(left[0]) && right[0].is_variable()) {
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/formula_preprocess.cpp:1692]: This should be `else if`, you are checking for `if(left.size() == 1 && right.size() == 1)` twice.

Comment (jurajsic) [reply src/smt/theory_str_noodler/formula_preprocess.cpp:1692]: Or maybe not `else if`, just another `if`.

File: src/smt/theory_str_noodler/counter_automaton.h [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+            output << "}" << std::endl;
+        }
+    };
+
+    
+}
+
+
+#endif
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/counter_automaton.h:178]:
```suggestion
#endif

```

File: src/smt/theory_str_noodler/counter_automaton.h [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+
+    using CounterAlphabet = StructAlphabet<std::set<AtomicSymbol>>;
+
+    /**
+     * @brief Counter Automaton
+     */
+    struct TagAut {
+        // underlying nfa
+        mata::nfa::Nfa nfa;
+        // structured alphabed with registe updates
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/counter_automaton.h:132]: `register`?

File: src/smt/theory_str_noodler/ca_str_constr.h [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+     * 
+     * @param not_conts Not contains
+     * @param autass Automata assignmnent after stabilization
+     * @return std::pair<LenNode, LenNodePrecision> LIA formula describing lengths of string models together with the precision.
+     */
+    std::pair<LenNode, LenNodePrecision> get_lia_for_not_contains(const Formula& not_conts, const AutAssignment& autass);
+
+}
+
+#endif
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/ca_str_constr.h:170]:
```suggestion
#endif

```

File: src/smt/theory_str_noodler/expr_solver.h (hunk 15 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/expr_solver.h]: Good point. This was a rebase artefact.

File: src/smt/theory_str_noodler/formula_preprocess.cpp (hunk 16 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/formula_preprocess.cpp]: Fixed

File: src/smt/theory_str_noodler/counter_automaton.h (hunk 18 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/counter_automaton.h]: Typo fixed

File: src/smt/theory_str_noodler/counter_automaton.h [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+    };
+
+    /**
+     * @brief Symbols of the form <mark, var, label, symbol>
+     */
+    struct AtomicSymbol {
+        char mark; // 0 = L; 1 = P; 2 = R
+        BasicTerm var; // variable from string constraint
+        char label; // 0 = missing value
+        mata::Symbol symbol; // symbol (used for the R-case)
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/counter_automaton.h:82]: What is the value if mark != R?

File: src/smt/theory_str_noodler/counter_automaton.h [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+            } else if (cmp > 0) {
+                return false;
+            } else {
+                cmp = label <=> as.label;
+                if(cmp < 0) {
+                    return true;
+                } else if (cmp > 0) {
+                    return false;
+                } else {
+                    cmp = symbol <=> as.symbol;
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/counter_automaton.h:97]: Does this comparison makes sense if mark is not R?

File: src/smt/theory_str_noodler/counter_automaton.h [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+            for (const auto& [key,val] : this->alph_symb_mata) {
+                ret.insert(key);
+            }
+            return ret;
+        }
+
+    };
+
+    /**
+     * @brief Symbols of the form <mark, var, label, symbol>
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/counter_automaton.h:76]: I would put some more explanation what mark,var,label means

File: src/smt/theory_str_noodler/parikh_image.h [hunk 23]
```diff
@@ -0,0 +1,217 @@
...
+#include <mata/nfa/strings.hh>
+#include <mata/nfa/builder.hh>
+
+#include "formula.h"
+#include "counter_automaton.h"
+#include "formula_preprocess.h"
+
+namespace smt::noodler::parikh {
+
+using Transition = std::tuple<mata::nfa::State, mata::Symbol, mata::nfa::State>;
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/parikh_image.h:23]: Wouldn't it better to use mata's version?

File: src/smt/theory_str_noodler/parikh_image.h [hunk 24]
```diff
@@ -0,0 +1,217 @@
...
+using TransitionCol = std::vector<std::vector<Transition>>;
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/parikh_image.h:24]: What is `Col`?

File: src/smt/theory_str_noodler/parikh_image.h (hunk 23 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/parikh_image.h:23]: Mata doesn't allow to compare transitions, which is necessary here. If we add comparison to mata I can switch it.

File: src/smt/theory_str_noodler/parikh_image.h (hunk 24 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/parikh_image.h:24]: Collection

File: src/smt/theory_str_noodler/counter_automaton.h (hunk 20 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/counter_automaton.h:82]: It is not defined as such a symbol should not occur.

**Comment by vhavlena** [comment]
I have checked the notcontains bug and it is indeed solved by handling notcontains inside the preprocessor (there is a rule requiring that a variable has no other occurrences and it did not take into account variables in notcontains).

File: src/smt/theory_str_noodler/counter_automaton.h (hunk 21 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/counter_automaton.h:97]: So-far not. In the future there should appear some new `M` mark.

File: src/smt/theory_str_noodler/counter_automaton.h (hunk 20 above)
Comment (jurajsic) [reply src/smt/theory_str_noodler/counter_automaton.h:85]: I mean, mark can be L, so there will be some symbol, is it going to be undefined? Or does R-case means something else, not the value of the mark?

File: src/smt/theory_str_noodler/counter_automaton.h (hunk 21 above)
Comment (jurajsic) [reply src/smt/theory_str_noodler/counter_automaton.h:100]: If we have two atomic symbols, both with mark `P`, same label, you will compare symbols, which from what I understand, is undefined, are you sure you want to do that? Shouldn't you compare just the stuff that is valid for given mark?

File: src/smt/theory_str_noodler/parikh_image.h (hunk 23 above)
Comment (jurajsic) [reply src/smt/theory_str_noodler/parikh_image.h:23]: Parikh image is not used anywhere else right now, right? If not, maybe I would move the stuff related to Parikh image (probably just the files `parikh_image.h` and `parikh_image.cpp`) to its own PR and we can sort all this later.

**Comment by jurajsic** [comment]
> I have checked the notcontains bug and it is indeed solved by handling notcontains inside the preprocessor (there is a rule requiring that a variable has no other occurrences and it did not take into account variables in notcontains).

Should that rule be fixed then? It should check for the variables in not contains. Or at least add a note that it does not check not contains.

File: src/smt/theory_str_noodler/ca_str_constr.h [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+         * 
+         * @return mata::nfa::Nfa Big NFA
+         */
+        mata::nfa::Nfa union_matrix() const;
+
+        const std::vector<BasicTerm>& get_var_order() const {
+            return this->var_order;
+        }
+
+        void set_aut(size_t copy, size_t var, const mata::nfa::Nfa& aut, bool re_offset = false) {
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/ca_str_constr.h:89]: Call it `recompute_offset` not `re_offset`, I was wondering what regular expression are you talking about.

**Comment by vhavlena** [comment]
> > I have checked the notcontains bug and it is indeed solved by handling notcontains inside the preprocessor (there is a rule requiring that a variable has no other occurrences and it did not take into account variables in notcontains).
> 
> 
> 
> Should that rule be fixed then? It should check for the variables in not contains. Or at least add a note that it does not check not contains.

It's ok. The Preprocessor now takes into account all variables in the system including those in notcontains.

File: src/smt/theory_str_noodler/parikh_image.h (hunk 24 above)
Comment (jurajsic) [reply src/smt/theory_str_noodler/parikh_image.h:24]: That is a pretty bad name, but I cannot give you any suggestions, because even `TransitionCollection` is very unspecific and I have no idea what is it used for. Is there some better name based on how you are using it?

**Comment by jurajsic** [comment]
> > > I have checked the notcontains bug and it is indeed solved by handling notcontains inside the preprocessor (there is a rule requiring that a variable has no other occurrences and it did not take into account variables in notcontains).
> > 
> > 
> > Should that rule be fixed then? It should check for the variables in not contains. Or at least add a note that it does not check not contains.
> 
> It's ok. The Preprocessor now takes into account all variables in the system including those in notcontains.

Ah ok, I get it now, this PR fixed it.

File: src/smt/theory_str_noodler/parikh_image.h (hunk 23 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/parikh_image.h:23]: It does not make sense as the content of this PR depends heavily on Parikh image computation. After merge we can discuss what we could do with it. Putting it to mata is a bad idea imo.

Comment (jurajsic) [reply src/smt/theory_str_noodler/parikh_image.h:23]: Ah ok, from the description of PR it seemed that you didn't use Parikh image at the end.

Comment (jurajsic) [reply src/smt/theory_str_noodler/parikh_image.h:23]: And I only meant putting comparison of transitions in Mata, so you do not have to use `Transition`.

File: src/smt/theory_str_noodler/ca_str_constr.h [hunk 26]
```diff
@@ -0,0 +1,170 @@
...
+            }
+        }
+
+        const mata::nfa::Nfa& get_aut(size_t copy, size_t var) const {
+            return this->aut_matrix[copy][var];
+        }
+    };
+
+    /**
+     * @brief Class for Tag aut generation for diseqations.
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/ca_str_constr.h:102]: Disequations or just one disequation?

File: src/smt/theory_str_noodler/ca_str_constr.h [hunk 27]
```diff
@@ -0,0 +1,170 @@
...
+        ca::TagAut construct_tag_aut();
+
+        const DiseqAutMatrix& get_aut_matrix() const {
+            return this->aut_matrix;
+        }
+
+    };
+
+    /**
+     * @brief Get LIA formula for a single disequations. The LIA formula describes all length
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/ca_str_constr.h:150]: Should it be `single disequation`? Or is this for all disequations? Because the name of the function seems like it is for all disequations, it even takes `Formula`.

File: src/smt/theory_str_noodler/decision_procedure.cpp [hunk 28]
```diff
@@ -1166,22 +1173,101 @@ namespace smt::noodler {
...
-                for (auto const &eq_from_diseq : replace_disequality(dis_or_eq)) {
-                    equations.add_predicate(eq_from_diseq);
+                // if we solve diesquations using CA --> we store the disequations to be solved later on
+                if(!single_diseq && this->m_params.m_ca_constr) {
+                    init_ca_diseq(dis_or_eq);
+                    single_diseq = true;
+                } else {
+                    for (auto const &eq_from_diseq : replace_disequality(dis_or_eq)) {
+                        equations.add_predicate(eq_from_diseq);
+                    }
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/decision_procedure.cpp:1257]: How does this works? If we have let's say 5 disequations, one of them is handled trough CA, and others are handled by the way we did it before (trough `to_code`)?

Comment (vhavlena) [reply src/smt/theory_str_noodler/decision_procedure.cpp:1257]: Yes, Michal works on extension to arbitrary number of disequations.

Comment (jurajsic) [reply src/smt/theory_str_noodler/decision_procedure.cpp:1257]: Ok, I would just call the bool variable differently, because right now I thought it should be checking if we have one disequation. Maybe `some_diseq_handled_by_ca`?

File: src/smt/theory_str_noodler/parikh_image.h (hunk 23 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/parikh_image.h:23]: Yes, I am planning to do the transition comparison in mata.

File: src/smt/theory_str_noodler/parikh_image.h (hunk 24 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/parikh_image.h]: I renamed it

File: src/smt/theory_str_noodler/ca_str_constr.h (hunk 25 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/ca_str_constr.h]: Renamed

File: src/smt/theory_str_noodler/ca_str_constr.h (hunk 26 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/ca_str_constr.h]: I fixed the comment.

File: src/smt/theory_str_noodler/ca_str_constr.h (hunk 27 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/ca_str_constr.h]: So-far just a single disequation, but this function is supposed to solve multiple disequations at some point. So I leave the name in plural.

File: src/smt/theory_str_noodler/decision_procedure.cpp (hunk 28 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/decision_procedure.cpp]: Nice, fixed.

File: src/smt/theory_str_noodler/ca_str_constr.h (hunk 27 above)
Comment (jurajsic) [reply src/smt/theory_str_noodler/ca_str_constr.h]: The next sentence has also only 'diseqeuation'. I would also put TODO/NOTE or something, that it handles only one disequation (or can it already handle multiple, just it is used with only one?)

File: src/smt/theory_str_noodler/counter_automaton.h (hunk 21 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/counter_automaton.h:100]: I made it clear. I created constructors for creating particular symbols setting default values to the unused values. I want to have the comparison function as general as possible because the semantics of symbols may change.

File: src/smt/theory_str_noodler/ca_str_constr.h (hunk 27 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/ca_str_constr.h]: I added a TODO comment.

File: src/smt/theory_str_noodler/counter_automaton.h (hunk 20 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/counter_automaton.h:85]: I made it clear through the constructor functions setting default values to unused fields.

Comment (jurajsic) [reply src/smt/theory_str_noodler/counter_automaton.h:85]: If you want to make it even safer, you can make the default constructor private, so you can only create AtomicSymbol using one of those three functions.

**Review by jurajsic — APPROVED** [review]
I did not check `ca_str_constr.cpp` not `parikh_image.cpp`.

File: src/smt/theory_str_noodler/parikh_image.h (hunk 24 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/parikh_image.h]: I added a comment to make it clear.

File: src/smt/theory_str_noodler/counter_automaton.h (hunk 20 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/counter_automaton.h:85]: Good idea, I made the constructor private.
